src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultClaimUidExtractor.cs

```diff
@@ -22,42 +22,94 @@ public DefaultClaimUidExtractor(ObjectPool<AntiforgerySerializationContext> pool
                 return null;
             }
 
-            var uniqueIdentifierParameters = GetUniqueIdentifierParameters(claimsIdentity);
             var claimUidBytes = ComputeSha256(uniqueIdentifierParameters);
             return Convert.ToBase64String(claimUidBytes);
         }
 
-        // Internal for testing
-        internal static IEnumerable<string> GetUniqueIdentifierParameters(ClaimsIdentity claimsIdentity)
+        public static IList<string> GetUniqueIdentifierParameters(IEnumerable<ClaimsIdentity> claimsIdentities)
         {
-            var nameIdentifierClaim = claimsIdentity.FindFirst(
-                claim => string.Equals(ClaimTypes.NameIdentifier, claim.Type, StringComparison.Ordinal));
-            if (nameIdentifierClaim != null && !string.IsNullOrEmpty(nameIdentifierClaim.Value))
+            if (claimsIdentities == null)
             {
-                return new string[]
+                return null;
+            }
+
+            foreach (var identity in claimsIdentities)
+            {
+                if (!identity.IsAuthenticated)
+                {
+                    continue;
+                }
+
+                var subClaim = identity.FindFirst(
+                    claim => string.Equals("sub", claim.Type, StringComparison.Ordinal));
+                if (subClaim != null && !string.IsNullOrEmpty(subClaim.Value))
+                {
+                    return new string[]
+                    {
+                        subClaim.Type,
+                        subClaim.Value,
+                        subClaim.Issuer
+                    };
+                }
+
+                var nameIdentifierClaim = identity.FindFirst(
+                    claim => string.Equals(ClaimTypes.NameIdentifier, claim.Type, StringComparison.Ordinal));
+                if (nameIdentifierClaim != null && !string.IsNullOrEmpty(nameIdentifierClaim.Value))
                 {
-                    ClaimTypes.NameIdentifier,
-                    nameIdentifierClaim.Value
-                };
+                    return new string[]
+                    {
+                        nameIdentifierClaim.Type,
+                        nameIdentifierClaim.Value,
+                        nameIdentifierClaim.Issuer
+                    };
+                }
+
+                var upnClaim = identity.FindFirst(
+                    claim => string.Equals(ClaimTypes.Upn, claim.Type, StringComparison.Ordinal));
+                if (upnClaim != null && !string.IsNullOrEmpty(upnClaim.Value))
+                {
+                    return new string[]
+                    {
+                        upnClaim.Type,
+                        upnClaim.Value,
+                        upnClaim.Issuer
+                    };
+                }
+            }
+
+            // We do not understand any of the ClaimsIdentity instances, fallback on serializing all claims in every claims Identity.
+            var allClaims = new List<Claim>();
```

Comment:
Since we are adding claims from multiple identities https://github.com/aspnet/Antiforgery/pull/52#discussion_r52498650, this can't be done with a simple `ToList()`. I don't think this will be an improvement.